### PR TITLE
[InstCombine] Match range check pattern with SExt

### PR DESCRIPTION
**Pull request by MatzeB**
= Background

We optimize range check patterns like the following:
```
  %n_not_negative = icmp sge i32 %n, 0
  call void @llvm.assume(i1 %n_not_negative)
  %a = icmp sge i32 %x, 0
  %b = icmp slt i32 %x, %n
  %c = and i1 %a, %b
```
to a single unsigned comparison:
```
  %n_not_negative = icmp sge i32 %n, 0
  call void @llvm.assume(i1 %n_not_negative)
  %c = icmp ult i32 %x, %n
```

= Extended Pattern

This adds support for a variant of this pattern where the upper range is compared with a sign extended value:

```
  %n_not_negative = icmp sge i64 %n, 0
  call void @llvm.assume(i1 %n_not_negative)
  %x_sext = sext i32 %x to i64
  %a = icmp sge i32 %x, 0
  %b = icmp slt i64 %x_sext, %n
  %c = and i1 %a, %b
```
is now optimized to:
```
  %n_not_negative = icmp sge i64 %n, 0
  call void @llvm.assume(i1 %n_not_negative)
  %x_sext = sext i32 %x to i64
  %c = icmp ult i64 %x_sext, %n
```

Alive2: https://alive2.llvm.org/ce/z/XVuz9L

### CODE REVIEW

File: llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp [hunk 1]
```diff
@@ -695,14 +695,27 @@ Value *InstCombinerImpl::simplifyRangeCheck(ICmpInst *Cmp0, ICmpInst *Cmp1,
...
+  Value *Cmp1Op0 = Cmp1->getOperand(0);
+  Value *Cmp1Op1 = Cmp1->getOperand(1);
   Value *RangeEnd;
-  if (Cmp1->getOperand(0) == Input) {
+  if (Cmp1Op0 == Input) {
     // For the upper range compare we have: icmp x, n
-    RangeEnd = Cmp1->getOperand(1);
-  } else if (Cmp1->getOperand(1) == Input) {
+    RangeEnd = Cmp1Op1;
+  } else if (isa<SExtInst>(Cmp1Op0) &&
```
Comment (dtcxzyw) [inline llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp:704]: Use `m_SExtOrSelf`.

Comment (MatzeB) [reply llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp]: ok

**Review by dtcxzyw — COMMENTED** [review]
Looks like it also holds for zext: https://alive2.llvm.org/ce/z/vpYSiD

BTW, can you add a proof for `(icmp slt x, 0) | (icmp sgt sext(x), n) --> icmp ugt x, n`?

**Comment by MatzeB** [comment]
> Looks like it also holds for zext: https://alive2.llvm.org/ce/z/vpYSiD

uh oh, the tests should have been `and %n, 9223372036854775807`. While this holds for `and i64 %n, 2147483647` where all upper 32bits are zero, the C++ code checks for non-negative number (which is what we actually want) and then zext won't hold any longer.

Let me change the tests to use `llvm.assume` instead to make things clearer.

**Comment by MatzeB** [comment]
@dtcxzyw 

> BTW, can you add a proof for (icmp slt x, 0) | (icmp sgt sext(x), n) --> icmp ugt x, n?

What do you mean by "add a proof"? More alive2 examples?

**Comment by MatzeB** [comment]
I added all the unit test examples to the updated alive2 link

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**